lib/playlistGenerator/Sequence.js

```diff
@@ -199,6 +189,46 @@ Sequence.prototype.checkExpires = function (expires) {
     return removedChunks;
 };
 
+Sequence.prototype.checkMaxAllowedClips = function (maxClipsPerFlavor) {
+
+    var filesToDelete = [];
+
+    // calc number of clips that are empty and reduce from total
+    let numEmptyClips = _.reduce(this.clips, (count, clip) => {
+        return clip.isEmpty() ? count + 1 : count;
+    }, 0);
+
+    let clips2Remove = Math.max(this.clips.length - numEmptyClips - maxClipsPerFlavor, 0);
+
+    // check if number of non-emtpy clips passed maximum allowed
+    if (clips2Remove > 0) {
+        this.logger.warn(`Max allowed clips exceeded. [total clips: ${this.clips.length}, max allowed: (${maxClipsPerFlavor}), clips to remove: ${clips2Remove}]`);
+
+        _.each(this.clips, (c) => {
+            if (!c.isEmpty() && clips2Remove > 0) {
+                clips2Remove--;
+
+                // for each source in removed clip, remove all chunks until clip is empty
+                _.each(c.inner.sources, (s) => {
+                    // empty source from chunks
+                    let paths = s.removeClip();
```

Comment:
maybe filesToDeleteUpdate should be incapsulated by clip object?
this will make code reuse easy: clip.clear will do all the work and return paths